Fetch missing beatmap when importing replay

cdwcgt opened the pull request:
- [x] Depends on https://github.com/ppy/osu/pull/24846

Reimplementation for
- https://github.com/ppy/osu/pull/14044
-  https://github.com/ppy/osu/pull/22741

Discussed in 
- https://github.com/ppy/osu/discussions/14029

Redesigned based on [peppy's comment](https://github.com/ppy/osu/pull/24450#issuecomment-1683608382)

I'm a little confused about the adjustment of anchor and origin, so the design may not look good, so please let me know if you are not satisfied

which this pr do:
1. Don't fetch delete pending beatmapset which use by [realm filter](https://github.com/ppy/osu/pull/24450#discussion_r1282485788)
2. Post a progress notification that will show beatmap card image and some metadata with a download button.
3. detected if there are already a beatmapset match online id but no beatmap match the score hash, which will happen with outdated beatmapset.
4. re-import score if beatmap which match the score imported.

If you think the design is okay, just add localization then done.

![image](https://github.com/ppy/osu/assets/34775378/4b114a8d-13ad-4774-b1ff-eb861405e65e)



https://github.com/ppy/osu/assets/34775378/257f6d5b-a2d7-4e95-8f7d-0d6678d47455



<details>
<summary>outdated</summary>

Tested importing from Stable, but it's annoying when missing beatmap keep appearing.

I cannot use DI to get `IPerformFromScreenRunner` in ScoreImporter, so I pass it by `ScoreManager`, I don't know if there is a better solution.

Some rulesets may need beatmap information when parsing the score, like mania need key count, so I choose don't to parse the score in advance

Current behavior:
1. Import failed due to missing beatmap.
5. Save score stream into memory for reimport.
6. Fetch from online by md5 in score, if no beatmap provided, stream will dispose and no difference from the original.
7. Push `ReplayMissingBeatmapScreen`.
8. If `AutomaticallyDownloadWhenSpectating` is true, then auto download.
9. When beatmap downloaded complete, import the score again.

after score imported or `ReplayMissingBeatmapScreen` dispose will dispose the stream.

https://github.com/ppy/osu/assets/34775378/c0accf2e-aa38-49b9-8e1c-add5af6c49d7

</details>

cdwcgt committed:
fetch missing beatmap when import score with missing beatmap

cdwcgt committed:
retrun DefaultBeatmap when beatmapset already delete pending

reimplement https://github.com/ppy/osu/pull/22741

cdwcgt committed:
ensure dispose stream

cdwcgt committed:
extract method

cdwcgt committed:
ensure Performer not null

cdwcgt committed:
don't passing stream by exception

bdach commented on osu.Game/Screens/Import/ReplayMissingBeatmapScreen.cs:
```diff
@@ -0,0 +1,225 @@
+
+            if (beatmapSetInfo == null)
+            {
+                var onlineBeatmapRequest = new GetBeatmapSetRequest(beatmap.OnlineBeatmapSetID);
+
+                onlineBeatmapRequest.Success += res =>
+                {
+                    beatmapSetInfo = res;
+                    updateStatus();
+                };
+                api.Queue(onlineBeatmapRequest);
```
Is this _second_ online request necessary? When is this ever gonna be `null`?

cdwcgt replied:
`BeatmapSet` in `APIBeatmap` is nullable?
I don't know if it will be null so I keep this logic

bdach replied:
It is nullable only because `APIBeatmap` is a structure reused by many endpoints, some of which will not return the beatmap set data by default. The lookup endpoint will, though, so this null path is essentially dead code.

I think it would be best to remove it and reduce some of the complexity in here.

bdach commented on osu.Game/Screens/Import/ReplayMissingBeatmapScreen.cs:
```diff
@@ -0,0 +1,225 @@
+
+        private APIBeatmapSet? beatmapSetInfo;
+
+        public ReplayMissingBeatmapScreen(APIBeatmap beatmap, MemoryStream? scoreStream = null)
```
Why is `scoreStream` allowed to be null if nothing ever uses that?

cdwcgt replied:
becuase `archive.GetStream()` retrun a nullable stream, so I make this nullable.
I will remove it.

bdach commented on osu.Game/Screens/Import/ReplayMissingBeatmapScreen.cs:
```diff
@@ -0,0 +1,225 @@
+        {
+            if (changes?.InsertedIndices == null) return;
+
+            if (beatmapSetInfo == null) return;
```
Does this mean what I'm thinking it means, namely that potentially a beatmap import can be missed entirely if the subscription completes before the API request for the beatmap set does?

bdach replied:
You misunderstand. The scenario I'm attempting to describe is:

```
API thread			import thread
----------------------------------------------------------------
begins beatmap set fetch
				begins importing beatmsp set
				to local database
				
				finishes importing beatmap set
				to local database
				
				realm subscription fires
				
				beatmap set hasn't yet been
				fetched from online
				=> subscription does nothing

beatmap set fetched
```

But as [already stated above](https://github.com/ppy/osu/pull/24450#discussion_r1287690508) I'm 99% sure all of this "nullable beatmap set" stuff can just be removed with no repercussions.

cdwcgt replied:
you mean declare that APIBeatmap.BeatmapSet is not null?

bdach replied:
Yes.

cdwcgt committed:
scoreStream never been null

cdwcgt committed:
remove all memory stream dispose

cdwcgt commented on osu.Game/Scoring/ScoreImporter.cs:
```diff
@@ -54,12 +60,41 @@ public ScoreImporter(RulesetStore rulesets, Func<BeatmapManager> beatmaps, Stora
+                Performer.PerformFromScreen(screen => screen.Push(new ReplayMissingBeatmapScreen(res, stream)));
+            };
+
+            req.Failure += _ => stream.Dispose();
```
ok, thanks to tell me about this.

cdwcgt committed:
declare BeatmapSet is not null

cdwcgt committed:
remove useless api resolved

cdwcgt committed:
remove useless Using

cdwcgt committed:
beatmapSetInfo never be null

peppy commented:
Looking at the video alone, the UX of this is still pretty shocking. Notifications popping up all over the place, a fullscreen view which doesn't let the user know what's going on and hijacks the full game. I feel like the only way this is going to work well is using a custom notification which allows downloading the beatmap inline.

```
You do not have the required beatmap for this replay

[ some kind of beatmap panel / donwnload interface ]

```

The notification would not dismiss while the download is happening, and after the local availability changes, it would retry the import in the same notification.

I'm not even sure it's worth reviewing this code without getting the UX correct first.

cdwcgt commented:
> I'm not even sure it's worth reviewing this code without getting the UX correct first.

Then I think we can start with custom notifications, but it may lack design.
Especially it seems that the notification may not fit that much content
If so, I will close this pr, or just simply show the corresponding beatmap for the score.
decision is yours

cdwcgt committed:
Merge branch 'master' into missing-beatmap

cdwcgt committed:
post a notification instead a screen

cdwcgt committed:
use realm query

cdwcgt committed:
clean up

cdwcgt committed:
check beatmap hash before try to import

cdwcgt committed:
use realm Subscription instead of Beatmap Download Tracker

cdwcgt committed:
ensure dispose realmSubscription

peppy commented:
Just based on the videos, this is looking much much better than what you had.

I'd hope we can make the inline beatmap card display a bit nicer, but the flow looks decent.

bdach committed:
Fix grammar in comment

bdach committed:
Merge branch 'beatmap-card-nano' into missing-beatmap

bdach committed:
Merge branch 'get-reader-regression' into missing-beatmap

cdwcgt commented:
I will update code to use new card style.

bdach commented:
Please don't, I'm actively working on this branch.

bdach committed:
Redo nano beatmap card design to fit needs better

Wanting to use this inside notification, it turns out that the original
design did not work very well at such narrow widths, and additionally
the typical button setup borrowed from elsewhere resulted in teeny tiny
action buttons.

To that end, slim down the design (get rid of thumbnail, audio preview,
make expandable right side slimmer), as well as change the entire panel
so that it has only one action associated with it at all times, and
clicking the panel in any place triggers that action.

bdach committed:
Fix `OsuTestScene.CreateAPIBeatmapSet()` not backlinking set beatmaps to the set

bdach committed:
Add very basic visual tests for missing beatmap notification

The full-stack test using the whole 9 `OsuGameTest` yards is unusable
for rapid development. I don't really get how you could ever design
anything using it without tossing your computer out the window.

bdach committed:
Integrate nano beatmap card into the notification

bdach committed:
Reword and localise copy

bdach committed:
Make missing beatmap notification simple

Progress didn't work for several reasons:

- It was spinning when nothing was actually happening yet
  (especially egregious with autodownload off)

- It was blocking exits (as all progress notifications do)

- When actually going to download, two progress notifications would
  pertain to one thing

- It wasn't helping much with the actual implementation of score
  re-import, cancelling the progress notification would result in
  similarly jank UX of beatmap importing but not the score.

bdach committed:
Rename spectator-specific settings to more generic (with backwards migration)

bdach committed:
Use different copy when auto-downloading

bdach committed:
Auto-close notification after successful download

bdach commented:
Ok, well, I tried to do something with this branch...

Of note:

- Visual appearance redesigned to be _somewhat_ palatable
- UX redesigned to be _somewhat_ tolerable
- Fixed autodownload setting parasiting off spectator autodownload and instead made one common autodownload setting, with appropriate copy and automatic migration

Note that this is still like 50% jank because of the amount of notifications fired. In an ideal world you'd probably want to eliminate both the "score failed" notification and the "importing" notification. But neither is trivial as those notifications are nestled _deep_ into `RealmArchiveModelImporter` and attempting to yank _that_ out would probably either take an eldritch abomination the likes of which shall never be witnessed, or a 1k line rewrite of central classes. So I'll just leave it halfway jank for now, I've already spent too much time on this in comparison to how much time I actually wanted to put in.

@peppy Leaving this for you to check UI/UX. Not sure you'll like my decisions, so go ham.

peppy committed:
Merge branch 'master' into missing-beatmap

peppy commented:
@bdach on an initial pass my main question would be why you can't click the card to bring up the overlay. Is it a mobile concern?

bdach commented:
Which overlay are you referring to specifically? It's a bit unclear, sorry.

peppy commented:
When clicking the new nano beatmap card inside the notification, it always triggers download (via [this override](https://github.com/ppy/osu/blob/5fcd7363320485cf35d25715edaf057cf646f6db/osu.Game/Beatmaps/Drawables/Cards/BeatmapCardNano.cs#L153)). Feels a bit odd to me.

bdach commented:
Well the general thinking was that the card is too small for multiple actions and tiny buttons that master has there, so it should just have one always. And if it's gonna have one always, it should be the most convenient one, i.e. "download the map now", rather than "show the beatmap set overlay, then download the map".

Dunno if valid, I guess there's no *real* reason not to have it open the overlay first. But that does beg the question where would download go on the card (if anywhere at all).

peppy commented:
> But that does beg the question where would download go on the card (if anywhere at all).

Hmm, but the weird thing is the download button is still there (and even more weirdly, it's the only part of the card that *doesn't* have a "Download" tooltip):

![2023-09-19 04 44 18@2x](https://github.com/ppy/osu/assets/191335/fcd5b9b3-4f0d-4b8b-bba2-3794bb866fae)

Judging from your last reply, was the button still being there unintentional?

bdach commented:
I'll take another look tomorrow. The tooltip thing is not intentional but the icon generally was supposed to be where it is (it was no longer meant to really be a button itself, more like an indication what is gonna happen when the card is clicked anyway).

I'll see if I can instead have just the expanded right part be the download button, and have the left part open the beatmap overlay, since that seems to be the UX you're angling for.

bdach committed:
Tweak nano beatmap card UX further to meet expectations

bdach commented:
@peppy I _think_ things should be closer to expectations with the latest commit. If they aren't let me know, or feel free to take over, I guess. Whichever's easier.

peppy committed:
Add comment mentioning why hover is disabled on the notification type

peppy committed:
Adjust sizing slightly to remove need for `CollapsibleButtonContainerSlim`

peppy committed:
Fix log output using incorrect name

peppy committed:
Pass `ArchiveReader` instead of `Stream` to simplify resolution code

peppy committed:
Remove unused translations

peppy commented on osu.Game/Database/MissingBeatmapNotification.cs:
```diff
@@ -0,0 +1,105 @@
+            {
+                if (r.All<BeatmapSetInfo>().Any(s => !s.DeletePending && s.OnlineID == beatmapSetInfo.OnlineID))
+                {
+                    Text = NotificationsStrings.MismatchingBeatmapForReplay;
```
Isn't this going to get overwritten by the `Text=` in `LoadComplete`?

bdach replied:
Yeah good spot. Should be addressed by https://github.com/ppy/osu/pull/24450/commits/ed9039f60f31415089be387e9075e6439099ee58.

peppy commented:
In addition to the one issue I haven't fixed, please check my new commits to make sure you agree with them.

bdach committed:
Fix notification text sets overwriting each other

bdach commented:
Other changes look good, thanks 👍

cdwcgt commented:
maybe we need to not auto download if there are a beatmapset match the online id?
It may a local change or something player need to keep, auto download will override them.

bdach commented:
@cdwcgt I'm pretty sure your original version had that too so ¯\_(ツ)_/¯

I think it's fine as is.
